### ISSUE TITLE
True negative with ValidDefaultValueSniff and typehinted args

### ISSUE THREAD

**Issue by josephzidell** [squizlabs/PHP_CodeSniffer#2136]
Code:
```php
<?php
class C
{

	/**
	 * Function
	 *
	 * @param stdClass|null $t Optional param
	 * @param integer       $s Required param
	 *
	 * @return void
	 */
	public function f(stdClass $t = null, int $s)
	{
	}

}
```

Excepted output:
`Arguments with default values must be at the end of the argument list`

Actual Output:
None, it passes

If I remove the typehint, the correct error shows up.
```php
<?php
class C
{

	/**
	 * Function
	 *
	 * @param stdClass|null $t Optional param
	 * @param integer       $s Required param
	 *
	 * @return void
	 */
	public function f($t = null, int $s)
	{
	}

}
```

**Comment by josephzidell** [comment]
Using php 7.1.19

**Comment by jrfnl** [comment]
This is not a bug. A very detailed explanation of this - the different between optional parameters and type-hinted parameters with a default value - can be found in: https://github.com/squizlabs/PHP_CodeSniffer/issues/1092

**Comment by josephzidell** [comment]
@jrfnl Thanks for pointing that out.

That is all true and good for PHP < 7.1. Since 7.1 has `?Type $t` in indicate "accepts null", therefore the addition of `= null` changes the argument from required to optional.

**Comment by josephzidell** [comment]
To be clear, example 2 in https://github.com/squizlabs/PHP_CodeSniffer/issues/1092#issuecomment-236814299 should be valid in php < 7.1

**Comment by gsherwood** [comment]
> For 7.1 and beyond, it should throw an error.

I believe the nullable operator does not break backwards compatibility with the previous `= null` syntax, so this syntax should be valid for all PHP versions.

**Comment by josephzidell** [comment]
It's certainly valid PHP syntax. I feel this sniff should throw an error in
this situation on php >= 7.1

**Comment by gsherwood** [comment]
> I feel this sniff should throw an error in
> this situation on php >= 7.1

That's not the job of this sniff. A sniff that says "dont use this syntax, use the nullable operator instead" would be handy, but it doesn't exist.
